File: src/main/kotlin/com/bitso/cli/auth/CredentialStore.kt
package com.bitso.cli.auth

import com.bitso.cli.model.BitsoError
import com.bitso.cli.model.ErrorCategory
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.attribute.PosixFilePermission
import java.nio.file.attribute.PosixFilePermissions

/**
 * Credential precedence: CLI flag > env var > config file.
 *
 * Lazily loaded — never instantiated by public commands.
 * The config file lives at `~/.config/bitso/config.toml` (Unix)
 * or `%APPDATA%\bitso\config.toml` (Windows).
 * Directories created with 0700, file with 0600 on Unix.
 */
class CredentialStore(
    private val cliApiKey: String? = null,
    private val cliApiSecret: ByteArray? = null
) {
    private val configDir: Path = resolveConfigDir()
    private val configFile: Path = configDir.resolve("config.toml")

    data class Credentials(
        val apiKey: String,
        val apiSecret: ByteArray
    )

    /**
     * Loads credentials in precedence order, throwing if none found.
     */
    fun load(): Credentials {
        // 1. CLI flag
        if (cliApiKey != null && cliApiSecret != null) {
            return Credentials(cliApiKey, cliApiSecret.copyOf())
        }

        // 2. Environment variables
        val envKey = System.getenv("BITSO_API_KEY")
        val envSecret = System.getenv("BITSO_API_SECRET")
        if (!envKey.isNullOrBlank() && !envSecret.isNullOrBlank()) {
            return Credentials(envKey, envSecret.toByteArray(Charsets.UTF_8))
        }

        // 3. Config file
        if (Files.exists(configFile)) {
            val config = parseConfigFile()
            if (config != null) return config
        }

        throw CredentialMissingException()
    }

    /**
     * Writes credentials to config file. Creates directory with restricted permissions.
     */
    fun write(apiKey: String, apiSecret: ByteArray): Path {
        Files.createDirectories(configDir)

        try {
            val perms = PosixFilePermissions.fromString("rwx------")
            Files.setPosixFilePermissions(configDir, perms)
        } catch (_: UnsupportedOperationException) {
            // Windows — POSIX not supported, ACL via icacls is out of scope for MVP
        }

        val body = """
            [api]
            key = "$apiKey"
            secret = "${String(apiSecret, Charsets.UTF_8)}"
        """.trimIndent()

        Files.writeString(configFile, body)

        try {
            val perms = PosixFilePermissions.fromString("rw-------")
            Files.setPosixFilePermissions(configFile, perms)
        } catch (_: UnsupportedOperationException) {
            // Windows
        }

        return configFile
    }

    private fun parseConfigFile(): Credentials? {
        var apiKey: String? = null
        var apiSecret: String? = null

        val lines = Files.readAllLines(configFile)
        for (line in lines) {
            val trimmed = line.trim()
            when {
                trimmed.startsWith("key =") -> {
                    apiKey = trimmed.substringAfter("\"").substringBeforeLast("\"")
                }
                trimmed.startsWith("secret =") -> {
                    apiSecret = trimmed.substringAfter("\"").substringBeforeLast("\"")
                }
            }
        }

        return if (apiKey != null && apiSecret != null) {
            Credentials(apiKey, apiSecret!!.toByteArray(Charsets.UTF_8))
        } else null
    }

    companion object {
        fun resolveConfigDir(): Path {
            val home = System.getProperty("user.home")
            val osName = System.getProperty("os.name").lowercase()

            return if (osName.contains("win")) {
                // %APPDATA%\bitso or fallback to ~\AppData\Roaming\bitso
                val appData = System.getenv("APPDATA")
                    ?: Paths.get(home, "AppData", "Roaming").toString()
                Paths.get(appData, "bitso")
            } else {
                // XDG base dir or ~/.config/bitso
                val xdgConfig = System.getenv("XDG_CONFIG_HOME")
                if (xdgConfig != null) {
                    Paths.get(xdgConfig, "bitso")
                } else {
                    Paths.get(home, ".config", "bitso")
                }
            }
        }
    }
}

class CredentialMissingException : RuntimeException(
    "No Bitso API credentials found. Use 'bitso auth set --api-key <key> --api-secret-stdin' to configure."
)

/**
 * Reads the full content of stdin into a ByteArray, then zeroes the intermediate buffer.
 * The secret NEVER becomes a String.
 */
fun readSecretFromStdin(): ByteArray {
    val buffer = ByteArray(4096)
    val out = ByteArrayOutputStream()
    var n: Int
    while (System.`in`.read(buffer).also { n = it } != -1) {
        out.write(buffer, 0, n)
    }
    val result = out.toByteArray()
    java.util.Arrays.fill(buffer, 0.toByte())
    out.reset()
    return result
}

/**
 * Reads the full content of a file path into a ByteArray.
 */
fun readSecretFromFile(path: Path): ByteArray {
    return Files.readAllBytes(path)
}

File: src/credential-store.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as os from 'node:os';
import { parse as parseToml } from 'smol-toml';

export interface Credentials {
  apiKey: string;
  apiSecret: Buffer;
}

function resolveConfigDir(): string {
  const platform = process.platform;
  if (platform === 'win32') {
    const appData = process.env['APPDATA'] ?? path.join(os.homedir(), 'AppData', 'Roaming');
    return path.join(appData, 'bitso');
  }
  const xdgConfig = process.env['XDG_CONFIG_HOME'];
  if (xdgConfig) return path.join(xdgConfig, 'bitso');
  return path.join(os.homedir(), '.config', 'bitso');
}

const CONFIG_DIR = resolveConfigDir();
const CONFIG_FILE = path.join(CONFIG_DIR, 'config.toml');

export class CredentialMissingError extends Error {
  constructor() {
    super("No Bitso API credentials found. Use 'bitso auth set --api-key <key> --api-secret-stdin' to configure.");
  }
}

export function loadCredentials(cliApiKey?: string, cliApiSecret?: Buffer): Credentials {
  if (cliApiKey && cliApiSecret) {
    return { apiKey: cliApiKey, apiSecret: Buffer.from(cliApiSecret) };
  }

  const envKey = process.env['BITSO_API_KEY'];
  const envSecret = process.env['BITSO_API_SECRET'];
  if (envKey && envSecret) {
    return { apiKey: envKey, apiSecret: Buffer.from(envSecret, 'utf8') };
  }

  if (fs.existsSync(CONFIG_FILE)) {
    const creds = parseConfigFile();
    if (creds) return creds;
  }

  throw new CredentialMissingError();
}

function parseConfigFile(): Credentials | null {
  try {
    const content = fs.readFileSync(CONFIG_FILE, 'utf8');
    const parsed = parseToml(content) as Record<string, unknown>;
    const api = parsed['api'] as Record<string, unknown> | undefined;
    if (!api) return null;
    const key = api['key'];
    const secret = api['secret'];
    if (typeof key === 'string' && typeof secret === 'string' && key && secret) {
      return { apiKey: key, apiSecret: Buffer.from(secret, 'utf8') };
    }
    return null;
  } catch {
    return null;
  }
}

export function writeCredentials(apiKey: string, apiSecret: Buffer): string {
  fs.mkdirSync(CONFIG_DIR, { recursive: true });

  try {
    fs.chmodSync(CONFIG_DIR, 0o700);
  } catch {
    // Windows — POSIX permissions not supported
  }

  const body = `[api]\nkey = "${apiKey}"\nsecret = "${apiSecret.toString('utf8')}"`;
  fs.writeFileSync(CONFIG_FILE, body, 'utf8');

  try {
    fs.chmodSync(CONFIG_FILE, 0o600);
  } catch {
    // Windows
  }

  return CONFIG_FILE;
}

export async function readSecretFromStdin(): Promise<Buffer> {
  const chunks: Buffer[] = [];
  for await (const chunk of process.stdin) {
    chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk as string));
  }
  const result = Buffer.concat(chunks);
  const last = result[result.length - 1];
  if (last === 0x0a) {
    return Buffer.from(result.subarray(0, result.length - 1));
  }
  return Buffer.from(result);
}

export function readSecretFromFile(filePath: string): Buffer {
  return fs.readFileSync(filePath);
}
